### PR TITLE
[libc++] Remove potential 0-sized array in __compressed_pair_padding

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
In the spirit of #105865 - I don't think we can use a similar implementation as what we did there, but that's the same idea / motivation

File: libcxx/include/__memory/compressed_pair.h [hunk 1]
```diff
@@ -52,13 +52,16 @@ _LIBCPP_BEGIN_NAMESPACE_STD
...
 
-template <class _ToPad>
+template <class _ToPad,
+          bool _Empty = ((is_empty<_ToPad>::value && !__libcpp_is_final<_ToPad>::value) ||
+                         is_reference<_ToPad>::value || sizeof(_ToPad) - __datasizeof_v<_ToPad> == 0)>
 class __compressed_pair_padding {
-  char __padding_[((is_empty<_ToPad>::value && !__libcpp_is_final<_ToPad>::value) || is_reference<_ToPad>::value)
-                      ? 0
-                      : sizeof(_ToPad) - __datasizeof_v<_ToPad>];
+  char __padding_[sizeof(_ToPad) - __datasizeof_v<_ToPad>];
```
Comment (ldionne) [inline libcxx/include/__memory/compressed_pair.h:59]: Can't we reuse the same `__padding` struct we now have in `<string>`?

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/__memory/compressed_pair.h (hunk 1 above)
Comment (serge-sans-paille) [reply libcxx/include/__memory/compressed_pair.h]: indeed

**Comment by serge-sans-paille** [comment]
It's not clear to me if the issue here come from the CI system or from my patch :-/

**Review by shafik — COMMENTED** [review]
This needs a more detailed summary, e.g. what is the motivation for this change?

**Comment by ldionne** [comment]
Please rebase onto main to re-trigger CI. The CI instability should be resolved now.

**Comment by serge-sans-paille** [comment]
@shafik : I'm trying to have libcxx compile with `-Wzero-length-array`. With these two commits, we're done.

@ldionne : I undid the reuse of `__padding`. The motivation is two folds: 1st it would create an extra header dependency (moving `__padding` definition to another header so that it can be included both in `string` and `compressed_pair.h`) and second  `compressed_pair` is now scarcely used, so it's probably ok to keep related stuff in the same header (?).

**Review by ldionne — APPROVED** [review]

**Pull request merged**